src/jdk.incubator.concurrent/share/classes/jdk/incubator/concurrent/ScopedValue.java

```diff
@@ -0,0 +1,723 @@
+
+    /**
+     * Creates a binding for a scoped value and runs a
+     * value-returning operation with that {@link ScopedValue} bound to the value.
+     * @param key the ScopedValue to bind
+     * @param value the value to bind it to, can be {@code null}
+     * @param <T> the type of the ScopedValue
+     * @param <U> the type of the Result
+     * @param op the operation to call
+     * @return the result
+     * @throws Exception if the operation completes with an exception
+     */
+    public static <T, U> U where(ScopedValue<T> key, T value, Callable<U> op) throws Exception {
+        return where(key, value).call(op);
+    }
+
+    /**
+     * Creates a binding for a scoped value variable and runs an
+     * operation with that  {@link ScopedValue} bound to the value.
+     * @param key the {@link ScopedValue} to bind
+     * @param value the value to bind it to, can be {@code null}
+     * @param <T> the type of the {@link ScopedValue}
+     * @param op the operation to run
+     */
+    public static <T> void where(ScopedValue<T> key, T value, Runnable op) {
+        where(key, value).run(op);
+    }
+
+    private ScopedValue() {
+        this.hash = generateKey();
+    }
+
+    /**
+     * Creates a scoped value to refer to a value of type T.
+     *
+     * @param <T> the type of the scoped value.
+     * @return a scoped value
+     */
+    public static <T> ScopedValue<T> newInstance() {
+        return new ScopedValue<T>();
+    }
+
+    /**
+     * Returns the current thread's bound value for this scoped-value variable.
+     * @return the value of the scoped value
+     * @throws NoSuchElementException if the scoped value is not bound
+     */
+    @ForceInline
+    @SuppressWarnings("unchecked")
+    public T get() {
+        Object[] objects;
+        if ((objects = scopedValueCache()) != null) {
+            // This code should perhaps be in class Cache. We do it
+            // here because the generated code is small and fast and
+            // we really want it to be inlined in the caller.
+            int n = (hash & Cache.SLOT_MASK) * 2;
+            if (objects[n] == this) {
+                return (T)objects[n + 1];
+            }
+            n = ((hash >>> Cache.INDEX_BITS) & Cache.SLOT_MASK) * 2;
+            if (objects[n] == this) {
+                return (T)objects[n + 1];
+            }
+        }
+        return slowGet();
+    }
+
+    @SuppressWarnings("unchecked")
+    private T slowGet() {
+        var value = findBinding();
+        if (value == Snapshot.NIL) {
+            throw new NoSuchElementException();
+        }
+        Cache.put(this, value);
+        return (T)value;
+    }
+
+    /**
+     * {@return {@code true} if the scoped value is bound to a value}
+     */
+    public boolean isBound() {
+        // ??? Do we want to search cache for this? In most cases we don't expect
+        // this {@link ScopedValue} to be bound, so it's not worth it. But I may
+        // be wrong about that.
```

Comment:
We should make it switchable by a runtime parameter. We should optionally cache failures.
Add a microbenchmark for that.